### ISSUE TITLE
Tanh bijector's fldj/ildj don't preserve dtype when numpy backend is used

### ISSUE THREAD

**Issue by 0x0badc0de** [tensorflow/probability#861]
TensorFlow version: 2.1.0, probability: 0.9.0
Nightly versions expose the same behaviour (TF: 2.2.0-dev20200402, TFP: 0.10.0-dev20200402)

In scope of https://github.com/tensorflow/probability/pull/800 I've added `sinh` bijector based on `tanh` one and it resulted in test failures in `JohnsonSU` distribution test added in the same PR. Failures are in dtype checking of dist methods. It happens with numpy backend only.

When bijector is implemented as `inline` one, ildj starts returning expected type.

Code to reproduce:
```
import numpy as np
import tensorflow as tf
import tensorflow_probability as tfp

tfp = tfp.experimental.substrates.numpy

tanh_bijector = tfp.bijectors.tanh.Tanh()

inline_bijector = tfp.bijectors.inline.Inline(
  forward_fn=tf.tanh,
  inverse_fn=tf.atanh,
  forward_log_det_jacobian_fn=lambda x: 2. * (np.log(2.) - x - tf.math.softplus(-2. * x)),
  forward_min_event_ndims=0,
  is_increasing=lambda: True,
  validate_args=True,
  name='tanh'
)

x = np.float32(0.5)
print(f'x={x}, x.dtype={x.dtype}')

for name, bijector in zip(('tanh', 'inline'), (tanh_bijector, inline_bijector)):
    y = bijector.forward(x)
    print(f'{name}: y={y}, y.dtype={y.dtype}')
    y = bijector.forward_log_det_jacobian(x, 0)
    print(f'{name}: y={y}, y.dtype={y.dtype}')
    y = bijector.inverse_log_det_jacobian(x, 0)
    print(f'{name}: y={y}, y.dtype={y.dtype}')
```

Output:
```
x=0.5, x.dtype=float32
tanh: y=0.46211716532707214, y.dtype=float32
tanh: y=-0.24022901391655516, y.dtype=float64
tanh: y=0.28768208236882487, y.dtype=float64
inline: y=0.46211716532707214, y.dtype=<dtype: 'float32'>
inline: y=-0.24022901391655516, y.dtype=float64
inline: y=0.2876819968223572, y.dtype=float32
```

Output if `tfp = tfp.experimental.substrates.numpy` is commented:
```
x=0.5, x.dtype=float32
tanh: y=0.46211716532707214, y.dtype=<dtype: 'float32'>
tanh: y=-0.24022895097732544, y.dtype=<dtype: 'float32'>
tanh: y=0.2876819968223572, y.dtype=<dtype: 'float32'>
inline: y=0.46211716532707214, y.dtype=<dtype: 'float32'>
inline: y=-0.24022895097732544, y.dtype=<dtype: 'float32'>
inline: y=0.2876819968223572, y.dtype=<dtype: 'float32'>
```

**Comment by brianwa84** [comment]
There's already a SinhArcsinh bijector, is that what you wanted?

Numpy is pretty aggressive about upcasting things to 64-bit, I think you'd
be best served by just using 64-bit everywhere w/ numpy or, if this is a
test failure, adding numpy_tags=['notap'] to whatever test is breaking.

Brian Patton | Software Engineer | bjp@google.com

**Comment by 0x0badc0de** [comment]
I cannot see parameters combinations for SinhArcsinh so it provides pure Sinh...

The test in question is the same as `testSampleLikeArgsGetDistDType` for `distributions.normal` so I copied `notap` option as well. It still fails but it didn't fail on February when I've submitted PR.

Just ran my test and `normal_test` with commit https://github.com/tensorflow/probability/pull/800/commits/82e18343a71ac7dccd7b5dc0c103314336bbda4a and without. Both tests fail without the commit and pass otherwise. Looks like that's because I've tested with `bazel test tensorflow_probability/python/distributions/johnson_su_test` while `testing/run_tests.sh` logic is a bit more complicated.

But doesn't `notap` just hides implementation issues? E.g. if someone uses tfp with tf backend he can see both types preserving and consistent return types. But that's not the case with numpy. Yet it doesn't contradict docs though.

Basically this is the motivation for reporting this issue. If this behaviour is acceptable, happy for it to be closed.

Still, do you think commit 82e18343a71ac7dccd7b5dc0c103314336bbda4a should stay so numpy tests pass eventhough they are not run on CI or it would be better to undo?
